### PR TITLE
[FLINK-9215][resoucemanager] Reduce noise in SlotPool's log to avoid confusing

### PR DESCRIPTION
**Pull request by sihuazhou**
## What is the purpose of the change

This PR fix the log in `SlotPool#releaseSlot()` to avoid confusing.

## Brief change log

  - *Only log the exception when the log level is TRACE in SlothPool#releaseSlot()*
  - *dd double check in ZooKeeperLeaderRetrievalService#start() to avoid race condition*

## Verifying this change

This change is a trivial rework / code cleanup without any test coverage.

## Does this pull request potentially affect one of the following parts:

no

## Documentation

no

### CODE REVIEW

**Comment by zentol** [comment]
I think it is fine that exceptions are logged under debug, it's just that we use exceptions in the normal life-cycle. I'm wondering if we could replace these kind of exceptions with a special `ExpectedLifecycleException?`  or similar that overrides `printStackTrace` to only contain the message.

**Comment by sihuazhou** [comment]
@zentol Thanks for you reply, I prefer the way to introduce a `ExpectedLifecycleException`, I think maybe it could be used in some other places in the future. Addressing this...

File: flink-runtime/src/main/java/org/apache/flink/runtime/leaderretrieval/ZooKeeperLeaderRetrievalService.java [hunk 1]
```diff
@@ -100,6 +100,12 @@ public void start(LeaderRetrievalListener listener) throws Exception {
 		LOG.info("Starting ZooKeeperLeaderRetrievalService {}.", retrievalPath);
 
 		synchronized (lock) {
+			// A double check for the race condition that multi invocation on this method
+			// could pass the first checkState() and wait on the lock, then they can all reset
+			// the leaderListener.
+			Preconditions.checkState(leaderListener == null, "ZooKeeperLeaderRetrievalService can " +
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/leaderretrieval/ZooKeeperLeaderRetrievalService.java]: This change is unrelated?

Comment (sihuazhou) [reply flink-runtime/src/main/java/org/apache/flink/runtime/leaderretrieval/ZooKeeperLeaderRetrievalService.java]: Ah, this is a hotfix...I would like to pull it out into another new PR if you object it here...

Comment (sihuazhou) [reply flink-runtime/src/main/java/org/apache/flink/runtime/leaderretrieval/ZooKeeperLeaderRetrievalService.java]: @zentol what do you think?

**Comment by sihuazhou** [comment]
Hi @zentol , will this PR go into 1.5? I think the log seems to be a bit noise currently when we diagnosing the job.

**Comment by sihuazhou** [comment]
cc @tillrohrmann

**Comment by sihuazhou** [comment]
cc @StephanEwen

**Comment by tillrohrmann** [comment]
@sihuazhou I'm actually no longer sure whether sending exceptions as causes along with the RPC messages was the best decision. Exceptions signal an exceptional situation and should not occur in the happy case.

Therefore, I'm a bit hesitant to introduce this special `normal-life-cycle-exception` because they should not exist in the first place since it is the normal life cycle. Instead, I think we should log the exception where they occur and send a `String reason` which can be logged on the remote component. That way we should get rid of most of the exceptions being printed on `DEBUG`.

As a first step to get there, I think it would be enough to simply print the exception message in the log statement. That way we would get rid of the stack traces and don't have to introduce a special exception type. As a second step, we can refactor the RPC interfaces which have the `Exception cause` parameters. What do you think?

FLINK-9277 is related but not directly the same. The scope of this issue is to decrease the redundant messages which occur when releasing a shared slot starting from the leafs.

**Comment by sihuazhou** [comment]
@tillrohrmann I think would be enough to simply print the exception message in the log, if we will refactor the PRC as a second step. Will update the PR.

**Comment by sihuazhou** [comment]
@tillrohrmann I updated the PR, could you please have a look again?

**Comment by sihuazhou** [comment]
Could anyone have a look at this? A minor change that would only effects the log...absolutely won't cause any regression problem, but would be helpful when user diagnosing the job according to the log...

File: flink-runtime/src/main/java/org/apache/flink/runtime/jobgraph/JobVertex.java [hunk 2]
```diff
@@ -304,7 +305,8 @@ public int getMaxParallelism() {
 	/**
 	 * Sets the maximum parallelism for the task.
 	 *
-	 * @param maxParallelism The maximum parallelism to be set. must be between 1 and Short.MAX_VALUE.
+	 * @param maxParallelism The maximum parallelism to be set.
+	 *                       Must be between 1 and {@link KeyGroupRangeAssignment#UPPER_BOUND_MAX_PARALLELISM}.
```
Comment (dawidwys) [inline flink-runtime/src/main/java/org/apache/flink/runtime/jobgraph/JobVertex.java]: Unrelated change. Will remove while merging, if you are ok with it @sihuazhou .
